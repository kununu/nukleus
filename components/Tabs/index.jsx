import React, {Component, PropTypes} from 'react';

import styles from './index.scss';

import {clearfix} from '../index.scss';


export default class Tabs extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.element).isRequired,
    pathname: PropTypes.string.isRequired
  };

  getNewProps (item) {
    const {props} = item;

    // Depending on which link it is (from react-router, from react-server, simple link) we need to access the local pathname according to the respective API
    const hash = props.to ? props.to.hash : '';
    const localPathname = props.href || props.path || props.to.pathname;

    const newProps = `${localPathname}${hash}` === `${this.props.pathname}${hash}` ? {className: styles.active} : {};

    return newProps;
  }

  render () {
    const {items} = this.props;
    return (
      <ul className={`${styles.tabs} ${clearfix}`}>
        {items.map((item, key) => (
          <li
            key={key} // eslint-disable-line react/no-array-index-key
            className={items.length > 1 ? '' : styles.pointerDisabled}>
            {React.cloneElement(item, this.getNewProps(item))}
          </li>
        ))}
      </ul>
    );
  }
}
