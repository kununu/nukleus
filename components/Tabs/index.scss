@import '~kununu-theme/scss/base/variables';

.tabs {
  border-bottom: 1px solid $gray-base-10;
  width: 100%;
  padding: 0;
  margin: 0;
  list-style: none;
  letter-spacing: -.02em;

  li {
    position: relative;
    display: block;
    float: left;

    @media (min-width: $screen-sm-min) {
      &:not(:first-child) {
        padding-left: 20px;
      }
    }

    a {
      padding: 10px 0;
      color: $gray-base-70;
      position: relative;
      display: block;
      text-decoration: none;

      &.active {
        border-bottom: 2px solid $blue;
        color: $gray-base-90;
      }
    }
  }

  @media (max-width: $screen-sm-min) {
    display: table;

    li {
      padding: 0 5px;
      display: table-cell;
      text-align: center;
      float: none;

      a {
        padding: 10px 0;
        font-size: $font-size-small;
      }

      &.active {
        border-bottom: 3px solid $blue;

        a {
          font-weight: $font-weight-semibold;
          color: $text-color;
          border-bottom: 0;
        }
      }
    }
  }
}

.pointerDisabled {
  pointer-events: none;
}
