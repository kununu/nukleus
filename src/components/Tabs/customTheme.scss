@import '~@kununu/kununu-theme-v2/scss/base/variables';

.blockTabs {
  border-bottom: 0;

  @media (max-width: $screen-sm-min) {
    display: table;
  }

  .tabItem {
    padding: 0;
    border: 0;
    display: table-cell;
    text-align: center;

    @media (min-width: $screen-sm-min) {
      text-align: left;
      display: block;
      float: left;
    }
  }

  .tabLink {
    box-sizing: border-box;
    border-radius: 10px 10px 0 0;
    border: 1px solid $gray-base-20;
    border-bottom: 0;
    background: $panel-bg;
    color: $white;
    font-size: 10px;
    text-transform: uppercase;
    font-weight: 600;
    padding: 13px 5px 12px;
    position: relative;
    display: block;
    text-decoration: none;

    &.tabActive {
      border-top: 4px solid $blue;
      padding-top: 10px;
      border-bottom: 0;
      color: $text-color;
    }

    @media (min-width: $screen-xs-min) {
      padding-left: 10px;
      padding-right: 10px;
    }

    @media (min-width: $screen-sm-min) {
      background: $teal;
      font-size: $font-size-base;
      padding: 17px 25px;

      &.tabActive {
        padding-top: 17px;
        border-top: 1px solid $gray-base-20;
        border-bottom: 1px solid $white;
        background: $panel-bg;
      }
    }
  }
}
