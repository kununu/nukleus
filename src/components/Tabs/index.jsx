import React from 'react';
import PropTypes from 'prop-types';

import ThemeContext from 'utils/themeContext';
import themeable from 'utils/theming';

import sharedStyles from '../index.scss';

import styles from './index.scss';

export default class Tabs extends React.Component {
  static propTypes = {
    hash: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.element).isRequired,
    pathname: PropTypes.string.isRequired,
    theme: PropTypes.oneOf(['default', 'block']),
  };

  static defaultProps = {
    hash: '',
    theme: 'default',
  };

  getNewProps (item, theme) {
    const {props} = item;
    const {hash: rootHash, pathname} = this.props;
    const itemHash = props.to && props.to.hash ? props.to.hash : '';

    // Option to manually set tab active on TNP
    const isTabActive = props.active;

    // Depending on which link it is (from react-router, from react-server, simple link) we need to access the local pathname according to the respective API
    const localPathname = `${props.href || props.path || (props.to && props.to.pathname)}${itemHash}`;
    const actualPathname = `${pathname}${rootHash}`;
    const isActive = localPathname === actualPathname;

    return {
      className: theme(props.className, 'tabLink', `${isTabActive || isActive && 'tabActive'}`),
    };
  }

  render () {
    const {items, theme: type} = this.props;
    const styleName = `${type}Tabs`;

    return (
      <ThemeContext.Consumer>
        {(context) => {
          const theme = themeable({...sharedStyles, ...styles, ...context});

          return (
            <ul className={theme('clearfix', 'tabs', styleName)}>
              {items.map((item, key) => (
                <li
                  key={key} // eslint-disable-line react/no-array-index-key
                  className={theme('tabItem', `${items.length <= 1 ? 'tabPointerDisabled' : ''}`)}
                >
                  {React.cloneElement(item, this.getNewProps(item, theme))}
                </li>
              ))}
            </ul>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
