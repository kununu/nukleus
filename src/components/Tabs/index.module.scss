@import '~@kununu/kununu-theme-v2/scss/base/variables';

.tabs {
  width: 100%;
  padding: 0;
  margin: 0;
  list-style: none;
  border-bottom: 1px solid $gray-base-10;
}

.tabItem {
  position: relative;
  display: inline-block;
  padding: 0 5px;
  box-sizing: border-box;

  &:first-of-type {
    padding-left: 0;
  }
}

.tabPointerDisabled {
  pointer-events: none;
}

.defaultTabs {
  .tabItem {
    @media (min-width: $screen-sm-min) {
      padding-right: 20px;
    }
  }
}

.tabLink {
  padding: 10px 0;
  color: $gray-base-70;
  position: relative;
  display: block;
  text-decoration: none;

  &.tabActive {
    border-bottom: 2px solid $blue;
    color: $gray-base-90;
  }

  @media (max-width: $screen-sm-min) {
    font-size: $font-size-small;
  }
}

.tabActive {
  .tabLink {
    font-weight: $font-weight-semibold;
    color: $text-color;
    border-bottom: 0;
  }
}

.blockTabs {
  border-bottom: 0;

  @media (max-width: $screen-sm-min) {
    display: table;
  }

  .tabItem {
    padding: 0;
    border: 0;
    border-right: 1px solid $gray-base-20;
    display: table-cell;
    text-align: center;

    &:first-child {
      border-left: 1px solid $gray-base-20;
    }

    @media (min-width: $screen-sm-min) {
      text-align: left;
      display: block;
      float: left;
    }
  }

  .tabLink {
    box-sizing: border-box;
    border-top: 1px solid $gray-base-20;
    border-bottom: 1px solid $gray-base-20;
    background: $panel-bg;
    color: $text-color;
    font-size: $font-size-small;
    font-weight: $font-weight-semibold;
    padding: 13px 5px 12px;
    position: relative;
    display: block;
    text-decoration: none;

    &.tabActive {
      border-top: 4px solid $blue;
      padding-top: 10px;
      border-bottom: 0;
    }

    @media (min-width: $screen-xs-min) {
      padding-left: 10px;
      padding-right: 10px;
    } 

    @media (min-width: $screen-sm-min) {
      background: $gray-base-10;
      font-size: $font-size-base;
      padding: 17px 25px;

      &.tabActive {
        padding-top: 17px;
        border-top: 1px solid $gray-base-20;
        border-bottom: 1px solid $white;
        background: $panel-bg;
      }
    }
  }
}
