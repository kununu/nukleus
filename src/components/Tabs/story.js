import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {withInfo} from '@storybook/addon-info';

import Tabs from './index';

function updateUrl (e, path) {
  e.preventDefault();
  const queryString = window.parent.location.search;

  window.parent.history.pushState({}, 'new', `${path}${queryString}`);
  window.history.go('/');
}

storiesOf('Tabs', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator((story, context) => withInfo('The tabs component! Comes in two flavors - a minimal default theme and a block theme. The component supports link components such as react-router link as well as simple anchor tags shown here. You can also pass url hashs to the component via the hash prop.')(story)(context))
  .add('basics', () => (
    <div>
      <h3>
        Default tab theme
      </h3>
      <Tabs
        items={[
          <a
            onClick={e => updateUrl(e, '/')}
            href="/"
          >
            First Tab
          </a>,
          <a
            onClick={e => updateUrl(e, '/2')}
            href="/2"
          >
            Second Tab
          </a>,
          <a
            onClick={e => updateUrl(e, '/3')}
            href="/3"
          >
            Third Tab
          </a>,
        ]}
        pathname={window.parent.location.pathname}
        theme="default"
      />

      <br />
      <br />
      <br />

      <h3>
        Block tab theme
      </h3>
      <Tabs
        items={[
          <a
            onClick={e => updateUrl(e, '/')}
            href="/"
          >
            First Tab
          </a>,
          <a
            onClick={e => updateUrl(e, '/2')}
            href="/2"
          >
            Second Tab
          </a>,
          <a
            onClick={e => updateUrl(e, '/3')}
            href="/3"
          >
            Third Tab
          </a>,
        ]}
        pathname={window.parent.location.pathname}
        theme="block"
      />
    </div>
  ));
