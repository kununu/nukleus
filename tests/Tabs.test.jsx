import React from 'react';
import renderer from 'react-test-renderer';
import Tabs from 'Tabs'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

test('Renders Tabs without crashing', () => {
  const component = renderer.create(
    <Tabs
      pages={[
        {path: 'test', query: 'test', title: 'First tab'},
        {path: '/test/2', query: 'test', title: 'Second tab'},
        {path: '/test/3', query: 'test', title: 'Third tab'}
      ]}
      pathname={'test'} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
