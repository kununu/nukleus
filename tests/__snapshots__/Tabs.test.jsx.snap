exports[`test Renders Tabs without crashing 1`] = `
<ul
  className="tabs clearfix">
  <li
    className="">
    <a
      className="active"
      onClick={[Function]}
      style={Object {}}>
      First tab
    </a>
  </li>
  <li
    className="">
    <a
      className={false}
      onClick={[Function]}
      style={Object {}}>
      Second tab
    </a>
  </li>
  <li
    className="">
    <a
      className={false}
      onClick={[Function]}
      style={Object {}}>
      Third tab
    </a>
  </li>
</ul>
`;
